webapp/app/.server/service/l2-cancellations.ts

```diff
@@ -239,3 +277,42 @@ export function getL2GovernorAddress(proposalType: L2CancellationType) {
   }
   return l2GovernorAddress;
 }
+
+export async function getAndUpdateL2CancellationByExternalId(
+  externalId: Hex
+): Promise<InferSelectModel<typeof l2CancellationsTable>> {
+  const proposal = await getL2CancellationByExternalId(externalId);
+  if (!proposal) {
+    throw notFound();
+  }
+
+  return await upgradeCancellationStatus(proposal, await getL2VetoNonce());
+}
+
+async function upgradeCancellationStatus(
+  cancellation: InferSelectModel<typeof l2CancellationsTable>,
+  currentNonce: number
+): Promise<InferSelectModel<typeof l2CancellationsTable>> {
+  if (cancellation.status !== l2CancellationStatusEnum.enum.ACTIVE) {
+    return cancellation;
+  }
+
+  if (hexSchema.safeParse(cancellation.transactionHash).success) {
+    cancellation.status = l2CancellationStatusEnum.enum.DONE;
+    await updateL2Cancellation(cancellation.id, cancellation);
+    return cancellation;
+  }
+
+  const state = await getL2ProposalState(cancellation.type, cancellation.externalId);
+  if (!isValidCancellationState(state)) {
+    cancellation.status = l2CancellationStatusEnum.enum.EXPIRED;
+    await updateL2Cancellation(cancellation.id, cancellation);
+  }
```

Comment:
At this point, do we need to keep going or can we return early?

Reply:
Lets return early. Otherwise it could become confusing